src/tool/subcommands/api_cmd/test_snapshot.rs

```diff
@@ -0,0 +1,172 @@
+    let genesis_header =
+        read_genesis_header(None, chain_config.genesis_bytes(&db).await?.as_deref(), &db).await?;
+    let chain_store = Arc::new(
+        ChainStore::new(
+            db.clone(),
+            db.clone(),
+            db.clone(),
+            chain_config.clone(),
+            genesis_header.clone(),
+        )
+        .unwrap(),
+    );
+    chain_store.set_heaviest_tipset(db.heaviest_tipset()?.into())?;
+    let state_manager =
+        Arc::new(StateManager::new(chain_store.clone(), chain_config, sync_config).unwrap());
+    let network_name = get_network_name_from_genesis(&genesis_header, &state_manager)?;
+    let message_pool = MessagePool::new(
+        MpoolRpcProvider::new(chain_store.publisher().clone(), state_manager.clone()),
+        network_name.clone(),
+        network_send.clone(),
+        Default::default(),
+        state_manager.chain_config().clone(),
+        &mut JoinSet::new(),
+    )?;
+
+    let peer_manager = Arc::new(PeerManager::default());
+    let sync_network_context =
+        SyncNetworkContext::new(network_send, peer_manager, state_manager.blockstore_owned());
+    let (shutdown, shutdown_recv) = mpsc::channel(1);
+    let rpc_state = Arc::new(RPCState {
+        state_manager,
+        keystore: Arc::new(tokio::sync::RwLock::new(KeyStore::new(
+            KeyStoreConfig::Memory,
+        )?)),
+        mpool: Arc::new(message_pool),
+        bad_blocks: Default::default(),
+        sync_state: Arc::new(RwLock::new(Default::default())),
+        eth_event_handler: Arc::new(EthEventHandler::new()),
+        sync_network_context,
+        network_name,
+        start_time: chrono::Utc::now(),
+        shutdown,
+        tipset_send,
+    });
+    rpc_state.sync_state.write().set_stage(SyncStage::Idle);
+    Ok((rpc_state, network_rx, shutdown_recv))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::daemon::db_util::download_to;
+    use itertools::Itertools as _;
+    use url::Url;
+
+    #[tokio::test]
+    async fn rpc_regression_tests() {
+        let urls = include_str!("test_snapshots.txt")
+            .trim()
+            .split("\n")
+            .filter_map(|n| {
+                Url::parse(
+                    format!("https://forest-snapshots.fra1.digitaloceanspaces.com/rpc_test/{n}")
+                        .as_str(),
+                )
+                .ok()
+            })
+            .collect_vec();
+        for url in urls {
+            print!("Testing {url} ...");
+            let tmp_dir = tempfile::tempdir().unwrap();
+            let tmp = tempfile::NamedTempFile::new_in(&tmp_dir)
+                .unwrap()
+                .into_temp_path();
+            println!("start downloading at {}", tmp.display());
+            download_to(&url, &tmp).await.unwrap();
+            println!("done downloading {}", tmp.display());
+            run_test_from_snapshot(&tmp).await.unwrap();
+            println!("  succeeded.");
+        }
```

Comment:
If I understand correctly, each test invocation will temporarily download the snapshots. The total snapshot size currently is around 4 MB, but I can easily see it going higher once we add more snapshots. This will be a pain for someone on a worse connection just trying to run tests.

Perhaps we could add something like `/.local/share/forest-dev/rpc-regression-snapshots` persistent path (it would vary from system to system, we could re-use the logic we have for the DB).

Reply:
Cache implemented.